compiler/rustc_typeck/src/check/method/suggest.rs

```diff
@@ -1035,16 +1035,29 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     // that had unsatisfied trait bounds
                     if unsatisfied_predicates.is_empty() {
                         let def_kind = lev_candidate.kind.as_def_kind();
-                        err.span_suggestion(
-                            span,
-                            &format!(
-                                "there is {} {} with a similar name",
-                                def_kind.article(),
-                                def_kind.descr(lev_candidate.def_id),
-                            ),
-                            lev_candidate.name,
-                            Applicability::MaybeIncorrect,
-                        );
+                        // Methods are defined within the context of a struct and their first parameter is always self,
+                        // which represents the instance of the struct the method is being called on
+                        // Associated functions don’t take self as a parameter and
+                        // they are not methods because they don’t have an instance of the struct to work with.
+                        if def_kind == DefKind::AssocFn && lev_candidate.fn_has_self_parameter {
+                            err.span_suggestion(
+                                span,
+                                &format!("there is a method with a similar name",),
+                                lev_candidate.name,
+                                Applicability::MaybeIncorrect,
+                            );
+                        } else {
+                            err.span_suggestion(
+                                span,
+                                &format!(
+                                    "there is {} {} with a similar name",
+                                    def_kind.article(),
+                                    def_kind.descr(lev_candidate.def_id),
```

Comment:
Wouldn’t it be better to adjust `descr` and `article` to do the right thing? It would most likely end up positively impacting a fair number of other diagnostics as well, and overall more consistent.

There are two ways this could be achieved done. A simpler, but with unclear performance impact, is to adjust `DefKind::AssocFn` to become an `DefKind::AssocFn { has_self: bool }`. The unfortunate implication of doing so is that this field would end up needing to be populated unconditionally, even if the only use for this data is in diagnostics. The second and probably safer perf-wise way would be to adjust `descr` to accept as an argument `TyCtxt` and use the [`associated_item` query](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.associated_item) with the `def_id` to obtain the relevant data.